core/deployment/src/main/java/io/quarkus/deployment/pkg/PackageConfig.java

```diff
@@ -56,6 +57,37 @@
     @ConfigItem
     public Optional<List<String>> userConfiguredIgnoredEntries;
 
+    /**
+     * List of all the dependencies that have been defined as optional to include into the final package of the application.
+     * Each optional dependency needs to be expressed in the following format:
+     * <p>
+     * groupId:artifactId:classifier:type
+     * <p>
+     * With the classifier and type being optional.
+     * <p>
+     * If the type is missing, the artifact is assumed to be of type {@code jar}.
+     * <p>
+     * This parameter is optional, if absent, no optional dependencies will be included into the final package of
+     * the application.
+     * <p>
+     * For backward compatibility reasons, this parameter is ignored by default and can be enabled by setting the
+     * parameter {@code quarkus.package.filter-optional-dependencies} to {@code true}.
+     * <p>
+     * This parameter is meant to be used in modules where multi-builds have been configured to avoid getting a final
+     * package with unused dependencies.
+     */
+    @ConfigItem
+    public Optional<Set<String>> optionalDependencies;
+
+    /**
+     * Flag indicating whether the optional dependencies should be filtered out or not.
+     * <p>
+     * This parameter is meant to be used in modules where multi-builds have been configured to avoid getting a final
+     * package with unused dependencies.
+     */
+    @ConfigItem(defaultValue = "false")
+    public boolean filterOptionalDependencies;
```

Comment:
Can you give a reasonable example of having configured `optionalDependencies` and wishing `filterOptionalDependencies` to be false?

Reply:
@aloubyansky No there is no such example. But I need a mechanism to enable the feature in order to distinguish, the case where we don't want to include any optional dependencies by not setting `optionalDependencies` intentionally and the case where `optionalDependencies` has not been set after migrating an application (what I called "backward compatibility reasons")

Reply:
It seems like there is some config redundancy. I.e. if you specify which dependencies you want to keep, it implies the rest of the optional deps you will want to be filtered out.

Reply:
I'm not sure, I get what you mean. This is only a flag to enable the feature.

Reply:
If you don't use this flag, there are several tests with optional dependencies that fail because it will filter out all the optional dependencies

Reply:
That's probably a sign of a bug? Why would all the optional dependencies be filtered out w/o it?
What I'm saying is if you configure `optionalDependencies`, you list the dependencies you want to keep. Which means the rest of the optional dependencies aren't desired, right?

Reply:
@aloubyansky No it is actually intended, let's say that you want to build you application twice for two different profiles foo and bar. The code for bar needs one additional library "ext-bar". So the dependency "ext-bar" will be marked as optional, foo won't defined anything for `includedOptionalDependencies` as it doesn't need any optional dependencies while bar will set  `ext-bar` as `includedOptionalDependencies`. Do you see what I mean?

Reply:
Yes, thanks. This PR actually enables reasonable use-cases for optional dependencies in Quarkus apps. Otherwise, I am not sure they make sense in Quarkus apps. There is a use-case of re-augmenting mutable-jar packaged apps though. But even then it might benefit from these config options.